Code/Framework/AzCore/AzCore/IO/Streamer/RecentlyUsedIndex.h

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#pragma once
+
+#include <AzCore/std/functional.h>
+#include <AzCore/std/numeric.h>
+#include <AzCore/std/smart_ptr/unique_ptr.h>
+#include <AzCore/std/typetraits/is_integral.h>
+
+namespace AZ::IO
+{
+    //! Uses an optimized double linked list to keep track of the most and least recently used index.
+    template<typename T>
+    class RecentlyUsedIndex
+    {
+        static_assert(
+            AZStd::is_integral_v<T> && AZStd::is_unsigned_v<T>,
+            "Recently Used Index only supports unsigned integers as they need to represent an index.");
+
+    public:
+        static constexpr T InvalidIndex = AZStd::numeric_limits<T>::max();
+
+        RecentlyUsedIndex() = default;
```

Comment:
Why do we even need this instead of using an existing data structure like a priority queue or something?  I'm kinda okay with it, but some of this seems like needless optimization, or needless templatization, or just, trying to make something new when std:: (especially since we now have x++20+ support) already provides.

Not saying to remove it, just surprised to see the move to custom container impl instead of using something that already exists.

Reply:
I don't think there's std container for LRU cache. The double list can be implemented with std::list but it's not as efficient in cache locality perspective.